README
NAME
    Dancer::Plugin::Auth::Extensible::Provider::DBIC - authenticate via
    DBIx::Class

VERSION
    version 0.1.2

DESCRIPTION
    This class is an authentication provider designed to authenticate users
    against a database using Dancer::Plugin::DBIC within the
    Dancer::Plugin::Auth::Extensible framework.

    See Dancer::Plugin::DBIC for how to configure a database connection
    appropriately; see the "CONFIGURATION" section below for how to
    configure this authentication provider with database details.

    See Dancer::Plugin::Auth::Extensible for details on how to use the
    authentication framework, including how to pick a more useful
    authentication provider.

CONFIGURATION
    This provider tries to use sensible defaults, so you may not need to
    provide much configuration if your database tables look similar to those
    in the "SUGGESTED SCHEMA" section below.

    The most basic configuration, assuming defaults for all options, and
    defining a single authentication realm named 'users':

        plugins:
            Auth::Extensible:
                realms:
                    users:
                        provider: 'DBIC'

    You would still need to have provided suitable database connection
    details to Dancer::Plugin::Database, of course; see the docs for that
    plugin for full details, but it could be as simple as, e.g.:

        plugins:
            Auth::Extensible:
                realms:
                    users:
                        provider: 'DBIC'
            DBIC:
                default:
                    dsn:          'dbi:SQLite:mydb.sqlite'
                    schema_class: My::Schema

    A full example showing all options:

        plugins:
            Auth::Extensible:
                realms:
                    users:
                        provider: 'DBIC'
                        # optionally set DB connection name to use (see named 
                        # connections in Dancer::Plugin::Database docs)
                        db_connection_name: 'foo'

                        # Optionally disable roles support, if you only want to check
                        # for successful logins but don't need to use role-based access:
                        disable_roles: 1

                        # optionally specify names of the table, relationship and
                        # columns (the values given below are the defaults)
                        users_resultset:     User
                        roles_relationship:  roles
                        role_column:         role
                        username_column:     username
                            # function in the user resultset 
                            # taking the submitted password
                            # and returning true if it matches
                            # the user's
                        password_check:      check_password

    See the main Dancer::Plugin::Auth::Extensible documentation for how to
    configure multiple authentication realms.

SUGGESTED SCHEMA
    If you use a DBIx::Class schema similar to the examples provided here,
    you should need minimal configuration to get this authentication
    provider to work for you.

  users class
        package My::Schema::Result::User;

        use strict;
        use warnings;

        use base qw/ DBIx::Class::Core /;

        __PACKAGE__->load_components(qw/EncodedColumn Core/);

        __PACKAGE__->table('Users');

        __PACKAGE__->add_columns(
            user_id => {
                data_type => 'INTEGER',
                is_auto_increment => 1,
                is_nullable => 0,
            },
            username => {
                data_type => 'VARCHAR',
                size => 32,
                is_nullable => 0,
            },
            password => {
                data_type => 'VARCHAR',
                size => 40,
                is_nullable => 0,
                encode_column => 1,
                encode_class  => 'Digest',
                encode_args   => { 
                    algorithm => 'SHA-1', 
                    format => 'hex',
                },
                encode_check_method => 'check_password',
            },
        );

        __PACKAGE__->set_primary_key( 'user_id' );
        __PACKAGE__->add_unique_constraint( 'username' => [ 'username' ] );

        __PACKAGE__->has_many( 
            user_roles => 'My::Schema::Result::UserRole', 'user_id' 
        );

        __PACKAGE__->many_to_many( 
            roles => 'user_roles', 'role'
        );

        1;

    You will quite likely want other fields to store e.g. the user object
    will be returned by the "logged_in_user" keyword for your convenience.

  roles class
    You'll need a table to store a list of available roles in (unless you're
    not using roles - in which case, disable role support (see the
    "CONFIGURATION" section).

        package My::Schema::Result::Role;

        use strict;
        use warnings;

        use base qw/ DBIx::Class::Core /;

        __PACKAGE__->table('Roles');

        __PACKAGE__->add_columns(
            role_id => {
                data_type => 'INTEGER',
                is_auto_increment => 1,
                is_nullable => 0,
            },
            role => {
                data_type => 'VARCHAR',
                size => 32,
                is_nullable => 0,
            },
        );

        __PACKAGE__->set_primary_key( 'role_id' );
        __PACKAGE__->add_unique_constraint( 'role' => [ 'role' ] );

        __PACKAGE__->has_many( 
            user_roles => 'My::Schema::Result::UserRole', 'role_id' 
        );

        __PACKAGE__->many_to_many( 
            users => 'user_roles', 'user'
        );

        1;

  user_roles class
    Finally, (unless you've disabled role support) you'll need a table to
    store user <-> role mappings (i.e. one row for every role a user has; so
    adding extra roles to a user consists of adding a new role to this
    table).

        package My::Schema::Result::UserRole;

        use strict;
        use warnings;

        use base qw/ DBIx::Class::Core /;

        __PACKAGE__->table('UserRoles');

        __PACKAGE__->add_columns(
            user_id => {
                data_type => 'INTEGER',
                is_foreign_key => 1,
                is_nullable => 0,
            },
            role_id => {
                data_type => 'INTEGER',
                is_foreign_key => 1,
                is_nullable => 0,
            },
        );

        __PACKAGE__->set_primary_key( 'user_id', 'role_id' );

        __PACKAGE__->belongs_to( 
            user => 'My::Schema::Result::User', 'user_id' 
        );
        __PACKAGE__->belongs_to( 
            role => 'My::Schema::Result::Role', 'role_id' 
        );

        1;

SEE ALSO
    Dancer::Plugin::Auth::Extensible

    Dancer::Plugin::DBIC

AUTHOR
    Yanick Champoux <yanick@cpan.org>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Yanick Champoux.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
Revision history for Dancer-Plugin-Auth-Extensible-DBIC

{{$NEXT}}
 [API CHANGES]

 [BUG FIXES]

 [DOCUMENTATION]

 [ENHANCEMENTS]

 [NEW FEATURES]

 [STATISTICS]

0.1.3 2013-08-23
 [BUG FIXES]
 - Remove dependency of test to DBIx::Class::EncodedColumn

 [STATISTICS]
 - code churn: 2 files changed, 14 insertions(+), 3 deletions(-)

0.1.2 2013-08-21
 [BUG FIXES]
 - Test::Compile is misbehaving...

 [STATISTICS]
 - code churn: 3 files changed, 11 insertions(+), 3 deletions(-)

0.1.1 2013-08-21
 [BUG FIXES]
 - Remove dependency on MyApp.

 [STATISTICS]
 - code churn: 2 files changed, 19 insertions(+), 1 deletion(-)

0.1.0 2013-08-21
 - First version, unleashed on an unsuspecting world.

 [STATISTICS]
 - code churn: 8 files changed, 120 insertions(+), 105 deletions(-)



t/dbic/MyApp.pm
package MyApp;

use Dancer ':syntax';
use Dancer::Plugin::DBIC;
use Crypt::SaltedHash;
use My::Schema;

# so that the config takes effect first
BEGIN { 
    set plugins => {
        DBIC => {
            default => {
                dsn => 'dbi:SQLite::memory:',
                schema_class => 'My::Schema',
            },
        },
        'Auth::Extensible' => {
            disable_roles => 0,
            realms => {
                users => {
                    provider        => 'DBIC',
                    users_resultset => 'Users',
                },
            },
        },
    };

    set session => 'Simple';

    set show_errors => 1;
    #set logger => 'console';
}

use Dancer::Plugin::Auth::Extensible;

get '/init' => sub {
    
    schema->deploy;

    my $user = rset('Users')->create({ username => 'bob', password => 'please' });

    my $role = rset('Roles')->create({ role => 'overlord' });

    $DB::single = 1;
    
    $user->add_to_roles($role, {});
};

get '/authenticate/:user/:password' => sub {
    my( $success, $realm ) = authenticate_user( ( map { param($_) } qw/ user password / ),
        'users' );

    if( $success ) {
        session logged_in_user => params->{user};
        session logged_in_user_realm => $realm;
    }

    return $success;
};

get '/roles' => sub {
    return join ':', user_roles;
};

1;


t/dbic/My/Schema/Result/Users.pm
package My::Schema::Result::Users;

use strict;
use warnings;

use base qw/ DBIx::Class::Core /;

__PACKAGE__->load_components(qw/EncodedColumn Core/);

__PACKAGE__->table('Users');

__PACKAGE__->add_columns(
    user_id => {
        data_type => 'INTEGER',
        is_auto_increment => 1,
        is_nullable => 0,
    },
    username => {
        data_type => 'VARCHAR',
        size => 32,
        is_nullable => 0,
    },
    secret => {
        data_type => 'VARCHAR',
        size => 40,
        is_nullable => 0,
        encode_column => 1,
        encode_class  => 'Digest',
        encode_args   => { 
            algorithm => 'SHA-1', 
            format => 'hex',
        },
        encode_check_method => 'check_secret',
    },
);

__PACKAGE__->set_primary_key( 'user_id' );
__PACKAGE__->add_unique_constraint( 'username' => [ 'username' ] );

__PACKAGE__->has_many( 
    user_roles => 'My::Schema::Result::UserRoles', 'user_id' 
);

__PACKAGE__->many_to_many( 
    roles => 'user_roles', 'role'
);

1;


Build.PL

use strict;
use warnings;

use Module::Build 0.3601;


my %module_build_args = (
  "build_requires" => {
    "Crypt::SaltedHash" => 0,
    "DBIx::Class::Core" => 0,
    "DBIx::Class::Schema" => 0,
    "Dancer::Plugin::Auth::Extensible" => 0,
    "Dancer::Test" => 0,
    "Module::Build" => "0.3601",
    "Test::More" => "0.88",
    "warnings" => 0
  },
  "configure_requires" => {
    "Module::Build" => "0.3601"
  },
  "dist_abstract" => "authenticate via DBIx::Class",
  "dist_author" => [
    "Yanick Champoux <yanick\@cpan.org>"
  ],
  "dist_name" => "Dancer-Plugin-Auth-Extensible-Provider-DBIC",
  "dist_version" => "0.1.2",
  "license" => "perl",
  "module_name" => "Dancer::Plugin::Auth::Extensible::Provider::DBIC",
  "recommends" => {},
  "recursive_test_files" => 1,
  "requires" => {
    "Dancer" => 0,
    "Dancer::Plugin::Auth::Extensible::Provider::Base" => 0,
    "Dancer::Plugin::DBIC" => 0,
    "base" => 0,
    "strict" => 0
  },
  "script_files" => []
);


my $build = Module::Build->new(%module_build_args);

$build->create_build_script;


lib/Dancer/Plugin/Auth/Extensible/Provider/DBIC.pm
package Dancer::Plugin::Auth::Extensible::Provider::DBIC;
# ABSTRACT: authenticate via DBIx::Class

use strict;
use base 'Dancer::Plugin::Auth::Extensible::Provider::Base';
use Dancer::Plugin::DBIC;
use Dancer qw(:syntax);


=head1 DESCRIPTION

This class is an authentication provider designed to authenticate users against
a database using L<Dancer::Plugin::DBIC> within the
L<Dancer::Plugin::Auth::Extensible> framework.

See L<Dancer::Plugin::DBIC> for how to configure a database connection
appropriately; see the L</CONFIGURATION> section below for how to configure this
authentication provider with database details.

See L<Dancer::Plugin::Auth::Extensible> for details on how to use the
authentication framework, including how to pick a more useful authentication
provider.


=head1 CONFIGURATION

This provider tries to use sensible defaults, so you may not need to provide
much configuration if your database tables look similar to those in the
L</SUGGESTED SCHEMA> section below.

The most basic configuration, assuming defaults for all options, and defining a
single authentication realm named 'users':

    plugins:
        Auth::Extensible:
            realms:
                users:
                    provider: 'DBIC'

You would still need to have provided suitable database connection details to
L<Dancer::Plugin::Database>, of course;  see the docs for that plugin for full
details, but it could be as simple as, e.g.:

    plugins:
        Auth::Extensible:
            realms:
                users:
                    provider: 'DBIC'
        DBIC:
            default:
                dsn:          'dbi:SQLite:mydb.sqlite'
                schema_class: My::Schema


A full example showing all options:

    plugins:
        Auth::Extensible:
            realms:
                users:
                    provider: 'DBIC'
                    # optionally set DB connection name to use (see named 
                    # connections in Dancer::Plugin::Database docs)
                    db_connection_name: 'foo'

                    # Optionally disable roles support, if you only want to check
                    # for successful logins but don't need to use role-based access:
                    disable_roles: 1

                    # optionally specify names of the table, relationship and
                    # columns (the values given below are the defaults)
                    users_resultset:     User
                    roles_relationship:  roles
                    role_column:         role
                    username_column:     username
                        # function in the user resultset 
                        # taking the submitted password
                        # and returning true if it matches
                        # the user's
                    password_check:      check_password

See the main L<Dancer::Plugin::Auth::Extensible> documentation for how to
configure multiple authentication realms.

=cut

=head1 SUGGESTED SCHEMA

If you use a DBIx::Class schema similar to the examples provided here, you should need
minimal configuration to get this authentication provider to work for you.

=head2 users class

    package My::Schema::Result::User;

    use strict;
    use warnings;

    use base qw/ DBIx::Class::Core /;

    __PACKAGE__->load_components(qw/EncodedColumn Core/);

    __PACKAGE__->table('Users');

    __PACKAGE__->add_columns(
        user_id => {
            data_type => 'INTEGER',
            is_auto_increment => 1,
            is_nullable => 0,
        },
        username => {
            data_type => 'VARCHAR',
            size => 32,
            is_nullable => 0,
        },
        password => {
            data_type => 'VARCHAR',
            size => 40,
            is_nullable => 0,
            encode_column => 1,
            encode_class  => 'Digest',
            encode_args   => { 
                algorithm => 'SHA-1', 
                format => 'hex',
            },
            encode_check_method => 'check_password',
        },
    );

    __PACKAGE__->set_primary_key( 'user_id' );
    __PACKAGE__->add_unique_constraint( 'username' => [ 'username' ] );

    __PACKAGE__->has_many( 
        user_roles => 'My::Schema::Result::UserRole', 'user_id' 
    );

    __PACKAGE__->many_to_many( 
        roles => 'user_roles', 'role'
    );

    1;

You will quite likely want other fields to store e.g. the user object
will be returned by the C<logged_in_user> keyword for your convenience.

=head2 roles class

You'll need a table to store a list of available roles in (unless you're not
using roles - in which case, disable role support (see the L</CONFIGURATION>
section).


    package My::Schema::Result::Role;

    use strict;
    use warnings;

    use base qw/ DBIx::Class::Core /;

    __PACKAGE__->table('Roles');

    __PACKAGE__->add_columns(
        role_id => {
            data_type => 'INTEGER',
            is_auto_increment => 1,
            is_nullable => 0,
        },
        role => {
            data_type => 'VARCHAR',
            size => 32,
            is_nullable => 0,
        },
    );

    __PACKAGE__->set_primary_key( 'role_id' );
    __PACKAGE__->add_unique_constraint( 'role' => [ 'role' ] );

    __PACKAGE__->has_many( 
        user_roles => 'My::Schema::Result::UserRole', 'role_id' 
    );

    __PACKAGE__->many_to_many( 
        users => 'user_roles', 'user'
    );

    1;

=head2 user_roles class

Finally, (unless you've disabled role support)  you'll need a table to store
user <-> role mappings (i.e. one row for every role a user has; so adding 
extra roles to a user consists of adding a new role to this table). 

    package My::Schema::Result::UserRole;

    use strict;
    use warnings;

    use base qw/ DBIx::Class::Core /;

    __PACKAGE__->table('UserRoles');

    __PACKAGE__->add_columns(
        user_id => {
            data_type => 'INTEGER',
            is_foreign_key => 1,
            is_nullable => 0,
        },
        role_id => {
            data_type => 'INTEGER',
            is_foreign_key => 1,
            is_nullable => 0,
        },
    );

    __PACKAGE__->set_primary_key( 'user_id', 'role_id' );

    __PACKAGE__->belongs_to( 
        user => 'My::Schema::Result::User', 'user_id' 
    );
    __PACKAGE__->belongs_to( 
        role => 'My::Schema::Result::Role', 'role_id' 
    );

    1;

=cut

sub authenticate_user {
    my ($self, $username, $password) = @_;

    # Look up the user:
    my $user = $self->get_user_details($username) or return;

    return $self->match_password($password, $user);
}

sub match_password {
    my( $self, $password, $user ) = @_;

    my $settings = $self->realm_settings;
    my $password_check = $settings->{password_check}  || 'check_password';

    return $user->$password_check( $password );
}


# Return details about the user.  The user's row in the users table will be
# fetched and all columns returned as a hashref.
sub get_user_details {
    my ($self, $username) = @_;

    return unless defined $username;

    my $settings = $self->realm_settings;

    # Get our database handle and find out the table and column names:
    my $database = schema($settings->{db_connection_name})
        or die "No database connection";

    my $users_table     = $settings->{users_resultset} || 'User';
    my $username_column = $settings->{username_column} || 'username';

    # Look up the user, 
    my $user = $database->resultset($users_table)->find({
            $username_column => $username
    }) or debug("No such user $username");

    return $user;
}

sub get_user_roles {
    my ($self, $username) = @_;

    my $settings = $self->realm_settings;

    # Get details of the user first; both to check they exist, and so we have
    # their ID to use.
    my $user = $self->get_user_details($username)
        or return;

    my $roles = $settings->{roles_relationship} || 'roles';
    my $role_column = $settings->{role_column} || 'role';

    return [ $user->$roles->get_column( $role_column )->all ];
}

1;

=head1 SEE ALSO

L<Dancer::Plugin::Auth::Extensible>

L<Dancer::Plugin::DBIC>

